### ISSUE TITLE
With a fresh container/volumes, dashboard is dead if SKIPGRAVITYONBOOT is set

### ISSUE THREAD

**Issue by DistractionRectangle** [pi-hole/docker-pi-hole#878]
This is a: Run Issue

## Details

The Web Interface shows a completely blank dashboard if gravity isn't initialized.
On bringing up a fresh container with empty volumes and `SKIPGRAVITYONBOOT` set, the dashboard fails to show any data. Upon updating gravity under `Tools -> Update Gravity` and navigating back to the dashboard, it becomes populated with data.

## Related Issues
- [X] I have searched this repository/Pi-hole forums for existing issues and pull requests that look similar 

I have not found an relevant issues in this repo, adminlte, main pihole repository or the forums; though I specifically searched for `SKIPGRAVITYONBOOT` as the keyword since it seems to be at the root of the issue.

## How to reproduce the issue 

1. Environment data
  * Operating System: **Ubuntu 20.04 LTS** 
  * Hardware: PC; AMD 3100, b550m itx
  * Kernel Architecture: x86/amd64
  * Docker Install Info and version: 
    - Software source: official docker-ce installed using their instructions
    - Supplimentary Software: NA
  * Hardware architecture: x86/amd64

2. docker-compose.yml contents, docker run shell command, or paste a screenshot of any UI based configuration of containers here

<details><summary><b>Erroneous docker-compose.yml</b></summary>
<p>

Command used to test docker-compose.yml changes
`docker-compose down && sudo rm -rf .docker-volumes/ && docker-compose up -d`

```YAML
services:
  pihole:
    cap_add:
    - NET_ADMIN
    environment:
      SKIPGRAVITYONBOOT: '1'
      ServerIP: 192.168.1.14
      WEBPASSWORD: password
    image: pihole/pihole:latest
    ports:
    - mode: host
      protocol: tcp
      published: 8080
      target: 80
    restart: unless-stopped
    volumes:
      - ./.docker-volumes/pihole/etc-dnsmasq.d:/etc/dnsmasq.d:rw
      - ./.docker-volumes/pihole/etc-pihole:/etc/pihole:rw
version: '3.7'
```

</p>
</details>

3. any additional info to help reproduce

## These common fixes ~~didn't~~ do work for my issue as detailed below. This is an edge case that can only be reached under specific circumstances, and more likely to be reached by new users since they'll have empty//uninitialized volumes.

- [X] I have tried removing/destroying my container, and re-creating a new container
- [X] I have tried fresh volume data by backing up and moving/removing the old volume data
- [X] I have tried running the stock `docker run` example(s) in the readme (removing any customizations I added)
- [ ] I have tried a newer or older version of Docker Pi-hole (depending what version the issue started in for me)
- [X] I have tried running without my volume data mounts to eliminate volumes as the cause

This is a minor issue, as manually initializing gravity does fix the issue, or not setting `SKIPGRAVITYONBOOT` on first run with a fresh instance//volume. 

The issue requires both that mounted volumes be empty//uninitialized and `SKIPGRAVITYONBOOT` set. Bringing up a container without persistent volumes but with `SKIPGRAVITYONBOOT` set functions as expected. Bringing up a container with empty persistence volumes but `SKIPGRAVITYONBOOT` not set also functions as intended. See these examples of functional configs that skirt the issue:

<details><summary><b>Working without volumes</b></summary>
<p>

```yaml
services:
  pihole:
    cap_add:
    - NET_ADMIN
    environment:
      SKIPGRAVITYONBOOT: '1'
      ServerIP: 192.168.1.14
      WEBPASSWORD: password
    image: pihole/pihole:latest
    ports:
    - mode: host
      protocol: tcp
      published: 8080
      target: 80
    restart: unless-stopped
#    volumes:
#      - ./.docker-volumes/pihole/etc-dnsmasq.d:/etc/dnsmasq.d:rw
#      - ./.docker-volumes/pihole/etc-pihole:/etc/pihole:rw
version: '3.7'
```

</p>
</details>

<details><summary><b>Working with out skipping gravity</b></summary>
<p>

```yaml
services:
  pihole:
    cap_add:
    - NET_ADMIN
    environment:
#      SKIPGRAVITYONBOOT: '1'
      ServerIP: 192.168.1.14
      WEBPASSWORD: password
    image: pihole/pihole:latest
    ports:
    - mode: host
      protocol: tcp
      published: 8080
      target: 80
    restart: unless-stopped
    volumes:
      - ./.docker-volumes/pihole/etc-dnsmasq.d:/etc/dnsmasq.d:rw
      - ./.docker-volumes/pihole/etc-pihole:/etc/pihole:rw
version: '3.7'
```

</p>
</details>

Edit: cut down the reproduction docker-compose files.

**Comment by DistractionRectangle** [comment]
Visual reference to the issue:
<details><summary>Blank Dashboard</summary>
<p>

![Dead_dash](https://user-images.githubusercontent.com/48034372/125869277-84512f67-6f3a-4aaa-beaf-60a5a9335dc3.png)

</p>
</details>
<details><summary>Updating Gravity</summary>
<p>

![Update_Gravity](https://user-images.githubusercontent.com/48034372/125869333-6e95e9aa-510f-4d73-aff1-2fabec4d7265.png)

</p>
</details>

<details><summary>Dashboard after updating Gravity</summary>
<p>

![Alive_Dash](https://user-images.githubusercontent.com/48034372/125869297-8437273b-6def-441e-b539-d8c5551e9e19.png)

</p>
</details>

**Comment by PromoFaux** [comment]
Interesting observation - thanks for the detailed report!

I will try to take a look after work today. In the mean time, if you are able - could you test this against the `:beta-v5.9` tag? That would at least rule out whether or not we've already inadvertently fixed the issue :)

**Comment by DistractionRectangle** [comment]
Tested with the beta-v5.9 tag earlier today and the issue is still present in that version. 

Far as I can tell, initializing the gravity db when none is found is delegated to [migrate_to_database](https://github.com/pi-hole/pi-hole/blob/master/gravity.sh#L276) in the gravity command, `pihole -g`.  `SKIPGRAVITYONBOOT` skips this whole command which allows the container to spin up into an uninitialized state.

Since this is a cross section of AdminLTE, pi-hole main and docker pi-hole, I'm not entirely sure where the appropriate place to apply a fix would be. One could do error handling in adminlte, or move `SKIPGRAVITYONBOOT` checks to gravity.sh in pi-hole main, or duplicate the initialization check into docker pi-hole. IMO, probably the cleanest way to do this would be to refactor the initialization check out of gravity.sh into it's own script, which both gravity and docker pi-hole could source.

**Comment by PromoFaux** [comment]
I think the appropriate place to fix is probably here in docker land. Error handling might also be nice on the bare metal stuff, but this situation comes about because of an environment variable over here.

From some testing, it only comes about when a volume is mounted to the container, and this makes sense because on first boot there will be nothing in that volume. (If one starts a container with no volume mapping then it works, but obvs then we have no persistence!)

I'm playing about with adding an initial check to `20-start.sh` that tests for the existence of `/etc/pihole/gravity.db`, and on the offchange file does not exist, then run gravity regardless of the `SKIPGRAVITYONBOOT` setting.

**Comment by DistractionRectangle** [comment]
Closed by #883
